File: sdks/python/src/glueco_sdk/transport.py
"""
Transport types and protocol for the Glueco Gateway SDK.

Defines GatewayTransport - the interface that plugins use to make requests.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, Generic, Iterator, Optional, TypeVar, Protocol, runtime_checkable


T = TypeVar("T")


@dataclass
class GatewayResponse(Generic[T]):
    """Response from a gateway request.
    
    Attributes:
        data: Parsed JSON response data.
        status: HTTP status code.
        headers: Response headers.
    """
    data: T
    status: int
    headers: Dict[str, str]


@dataclass
class GatewayStreamResponse:
    """Streaming response from a gateway request.
    
    Attributes:
        stream: Iterator of bytes chunks (SSE or raw).
        status: HTTP status code.
        headers: Response headers.
    """
    stream: Iterator[bytes]
    status: int
    headers: Dict[str, str]
    
    def iter_lines(self) -> Iterator[str]:
        """Iterate over SSE lines."""
        buffer = b""
        for chunk in self.stream:
            buffer += chunk
            while b"\n" in buffer:
                line, buffer = buffer.split(b"\n", 1)
                yield line.decode("utf-8").strip()


@runtime_checkable
class GatewayTransport(Protocol):
    """Protocol for gateway transport implementations.
    
    Plugins use this interface to make signed requests.
    The SDK provides implementations via create_transport().
    
    Example:
        >>> def my_plugin(transport: GatewayTransport):
        ...     response = transport.request(
        ...         resource_id="llm:groq",
        ...         action="chat.completions",
        ...         payload={...},
        ...     )
        ...     return response.data
    """
    
    def request(
        self,
        resource_id: str,
        action: str,
        payload: Dict[str, Any],
        *,
        method: str = "POST",
        headers: Optional[Dict[str, str]] = None,
        timeout: Optional[float] = None,
    ) -> GatewayResponse[Any]:
        """Make a signed request to a resource action."""
        ...
    
    def request_stream(
        self,
        resource_id: str,
        action: str,
        payload: Dict[str, Any],
        *,
        method: str = "POST",
        headers: Optional[Dict[str, str]] = None,
        timeout: Optional[float] = None,
    ) -> GatewayStreamResponse:
        """Make a streaming signed request."""
        ...
    
    @property
    def proxy_url(self) -> str:
        """Gateway proxy URL."""
        ...
    
    @property
    def app_id(self) -> str:
        """Application ID."""
        ...

File: packages/sdk/src/transport.ts
// ============================================
// GATEWAY TRANSPORT INTERFACE
// The signed-request interface createTransport() returns: PoP signing,
// base URL handling, and error parsing, behind a single typed
// request()/requestStream() call keyed by resourceId + action.
// ============================================

/**
 * Request options for gateway transport.
 */
export interface GatewayRequestOptions {
  /** HTTP method override (default: POST) */
  method?: "GET" | "POST" | "PUT" | "DELETE" | "PATCH";

  /** Custom headers to include */
  headers?: Record<string, string>;

  /** Request timeout in milliseconds */
  timeout?: number;

  /** AbortSignal for cancellation */
  signal?: AbortSignal;

  /** Enable streaming response */
  stream?: boolean;
}

/**
 * Response from gateway transport.
 */
export interface GatewayResponse<T = unknown> {
  /** Response data (for non-streaming) */
  data: T;

  /** Response status code */
  status: number;

  /** Response headers */
  headers: Record<string, string>;
}

/**
 * Streaming response from gateway transport.
 */
export interface GatewayStreamResponse {
  /** Readable stream of response data */
  stream: ReadableStream<Uint8Array>;

  /** Response status code */
  status: number;

  /** Response headers */
  headers: Record<string, string>;
}

/**
 * Gateway Transport Interface
 *
 * Returned by createTransport(). Abstracts away PoP signing, baseURL
 * handling, and error parsing behind a single typed call.
 *
 * @example
 * ```ts
 * const transport = createTransport({ proxyUrl, appId });
 * const { data } = await transport.request<ChatResponse>(
 *   "llm:gemini",
 *   "chat.completions",
 *   payload
 * );
 * ```
 */
export interface GatewayTransport {
  /**
   * Make a typed request to a resource action.
   *
   * @param resourceId - Resource identifier (e.g., "llm:gemini")
   * @param action - Action name (e.g., "chat.completions")
   * @param payload - Request payload (will be JSON serialized)
   * @param options - Optional request options
   * @returns Promise resolving to typed response
   */
  request<TResponse = unknown, TPayload = unknown>(
    resourceId: string,
    action: string,
    payload: TPayload,
    options?: GatewayRequestOptions,
  ): Promise<GatewayResponse<TResponse>>;

  /**
   * Make a streaming request to a resource action.
   *
   * @param resourceId - Resource identifier (e.g., "llm:gemini")
   * @param action - Action name (e.g., "chat.completions")
   * @param payload - Request payload (will be JSON serialized)
   * @param options - Optional request options
   * @returns Promise resolving to stream response
   */
  requestStream<TPayload = unknown>(
    resourceId: string,
    action: string,
    payload: TPayload,
    options?: Omit<GatewayRequestOptions, "stream">,
  ): Promise<GatewayStreamResponse>;

  /**
   * Get the base proxy URL.
   * Useful for constructing URLs for vendor SDKs.
   */
  getProxyUrl(): string;

  /**
   * Get the PoP-signed fetch function.
   * Use this when you need to use vendor SDKs that require a custom fetch.
   *
   * @returns Fetch function with PoP signing
   */
  getFetch(): typeof fetch;
}
